Quotient/e2ee/cryptoutils.cpp

```diff
@@ -183,30 +197,31 @@ SslExpected<QByteArray> Quotient::aesCtr256Decrypt(const QByteArray& ciphertext,
         return ERR_get_error();
     }
 
-    int length = 0;
-    int plaintextLength = 0;
-    const int ciphertextSize = checkedSize(ciphertext.size()).first;
-    auto decrypted = zeroedByteArray(ciphertext.size());
-
-    if (EVP_DecryptInit_ex(context.get(), EVP_aes_256_ctr(), nullptr, reinterpret_cast<const unsigned char *>(key.constData()), reinterpret_cast<const unsigned char *>(iv.constData())) != 1) {
-        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
-        return ERR_get_error();
+    const auto [ciphertextSize, clamped] = checkedSize(ciphertext.size());
+    if (clamped) {
+        qCCritical(E2EE) << "The ciphertext is too long to be decrypted";
+        Q_ASSERT(!clamped); // We just don't deal with such sizes in Matrix?
+        return SslPayloadTooLong;
     }
 
-    if (EVP_DecryptUpdate(context.get(), reinterpret_cast<unsigned char*>(decrypted.data()), &length, reinterpret_cast<const unsigned char*>(ciphertext.constData()), ciphertextSize) != 1) {
-        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
-        return ERR_get_error();
-    }
+    FixedBuffer<> decrypted(static_cast<size_t>(ciphertextSize),
+                            FixedBufferBase::FillWithZeros);
+    CALL_OPENSSL(EVP_DecryptInit_ex(context.get(), EVP_aes_256_ctr(), nullptr,
+                                    key.data(), iv.data()));
 
-    plaintextLength = length;
-    if (EVP_DecryptFinal_ex(context.get(), reinterpret_cast<unsigned char*>(decrypted.data()) + length, &length) != 1) {
-        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
-        return ERR_get_error();
-    }
+    int decryptedLength = 0;
+    CALL_OPENSSL(EVP_DecryptUpdate(
+        context.get(), decrypted.data(), &decryptedLength,
+        reinterpret_cast<const unsigned char*>(ciphertext.constData()),
+        ciphertextSize));
+
+    int tailLength = -1; // Normally becomes 0 after the next call
```

Comment:
what's "normally"?

Reply:
For the current cipher it can be an assertion actually (on both occasions with this comment).